common/asyncworkflow/queue/consumer/default_consumer.go

```diff
@@ -0,0 +1,262 @@
+	"github.com/uber/cadence/common/codec"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/messaging"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/types"
+)
+
+const (
+	defaultShutdownTimeout = 5 * time.Second
+	defaultConcurrency     = 100
+)
+
+type DefaultConsumer struct {
+	queueID         string
+	innerConsumer   messaging.Consumer
+	logger          log.Logger
+	scope           metrics.Scope
+	frontendClient  frontend.Client
+	ctx             context.Context
+	cancelFn        context.CancelFunc
+	wg              sync.WaitGroup
+	shutdownTimeout time.Duration
+	msgDecoder      codec.BinaryEncoder
+	concurrency     int
+}
+
+type Option func(*DefaultConsumer)
+
+func WithConcurrency(concurrency int) Option {
+	return func(c *DefaultConsumer) {
+		c.concurrency = concurrency
+	}
+}
+
+func New(
+	queueID string,
+	innerConsumer messaging.Consumer,
+	logger log.Logger,
+	metricsClient metrics.Client,
+	frontendClient frontend.Client,
+	options ...Option,
+) *DefaultConsumer {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	c := &DefaultConsumer{
+		queueID:         queueID,
+		innerConsumer:   innerConsumer,
+		logger:          logger.WithTags(tag.AsyncWFQueueID(queueID)),
+		scope:           metricsClient.Scope(metrics.AsyncWorkflowConsumerScope),
+		frontendClient:  frontendClient,
+		ctx:             ctx,
+		cancelFn:        cancelFn,
+		shutdownTimeout: defaultShutdownTimeout,
+		msgDecoder:      codec.NewThriftRWEncoder(),
+		concurrency:     defaultConcurrency,
+	}
+
+	for _, opt := range options {
+		opt(c)
+	}
+
+	return c
+}
+
+func (c *DefaultConsumer) Start() error {
+	if err := c.innerConsumer.Start(); err != nil {
+		return err
+	}
+
+	for i := 0; i < c.concurrency; i++ {
+		c.wg.Add(1)
+		go c.runProcessLoop()
+		c.logger.Info("Started process loop", tag.Counter(i))
+	}
+	c.logger.Info("Started consumer", tag.Dynamic("concurrency", c.concurrency))
+	return nil
+}
+
+func (c *DefaultConsumer) Stop() {
+	c.logger.Info("Stopping consumer")
+	c.cancelFn()
+	c.wg.Wait()
+	if !common.AwaitWaitGroup(&c.wg, c.shutdownTimeout) {
+		c.logger.Warn("Consumer timed out on shutdown", tag.Dynamic("timeout", c.shutdownTimeout))
+		return
+	}
+
+	c.innerConsumer.Stop()
+	c.logger.Info("Stopped consumer")
+}
+
+func (c *DefaultConsumer) runProcessLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case msg, ok := <-c.innerConsumer.Messages():
+			if !ok {
+				c.logger.Info("Consumer channel closed")
+				return
+			}
+
+			c.processMessage(msg)
+		case <-c.ctx.Done():
+			c.logger.Info("Consumer context done so terminating loop")
+			return
+		}
+	}
+}
+
+func (c *DefaultConsumer) processMessage(msg messaging.Message) {
+	logger := c.logger.WithTags(tag.Dynamic("partition", msg.Partition()), tag.Dynamic("offset", msg.Offset()))
+	logger.Debug("Received message")
+
+	sw := c.scope.StartTimer(metrics.ESProcessorProcessMsgLatency)
```

Comment:
should be `AsyncWorkflowProcessMsgLatency`

Reply:
good catch